### ISSUE TITLE
Haywire zero-copy

### ISSUE THREAD

**Issue by kellabyte** [haywire/haywire#75]
There is some zero-copying in Haywire but not enough. Now that I understand C more and libuv more I suspect we are copying memory when we do not need to in Haywire.

https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_server.c#L248
Here we allocate the memory libuv requests for the TCP reads on the heap and we should be able to use `hw_string` everywhere and never need to memcpy headers or any HTTP field if we free() this memory at the very end of the request.

Currently we free this buffer after the request callback is complete to the user code but what if the user spawns a thread? We need to offer a free API to users of the library. That way we never need to have any memcpy in Haywire.

Examples of memcpy that could be removed:
https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_request.c#L115    
https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_request.c#L141    
https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_request.c#L157    
https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_server.c#L248    

This should reduce some of our most expensive calls of malloc() and memcpy().

**Comment by kellabyte** [comment]
Currently Haywire automatically free's the `http_request` here https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_request.c#L285. It assumes that calling the users callback will be synchronous. I think we should move this to here https://github.com/kellabyte/Haywire/blob/master/src/haywire/http_server.c#L293 so that the user can spawn a thread. This would facilitate zero-copy as well.

**Comment by kellabyte** [comment]
Haywire already has a send_response_complete() callback. See the hello world sample.
```
void response_complete(void* user_data)
{
    // Here we would could free http_request and hw_http_response (should make names consistent)
}

void get_root(http_request* request, hw_http_response* response, void* user_data)
{
    // ... Do a bunch of stuff...

    // This could also free http_request.
    hw_http_response_send(response, "user_data", response_complete);
}
```

**Comment by kellabyte** [comment]
I think the current haywire free's buffers at the wrong spot. It free's them after parsing/calling the callback when it should do it after the response is sent.

**Comment by nmdguerreiro** [comment]
Fixed by #103
